obsplus/events/validate.py

```diff
@@ -100,57 +99,191 @@ def check_origins(event: Event):
+    """
+    Ensure there are no picks with the same phases on the same channels.
+    """
+    # first get dataframe of picks
+    pdf = obsplus.picks_to_df(event)
+    pdf = pdf.loc[pdf.evaluation_status != "rejected"]
+    event_id = str(event.resource_id)
+
+    def dup_picks(phase_hint, on="station"):
+        """ function for checking """
+        df = pdf.loc[pdf.phase_hint == phase_hint]
+        bad = df.loc[df[on].duplicated()][on].tolist()
+        assert len(bad) == 0, (
+            f"Duplicate {phase_hint} picks found\n"
+            f"event_id: {event_id}, "
+            f"{on}/s: {bad}"
+        )
+
+    # Checking for duplicated picks
+    dup_picks("P")
+    dup_picks("S")
+    dup_picks("IAML", on="seed_id")
+    dup_picks("AML", on="seed_id")
+
+
+@validator("obsplus", Event)
+def check_pick_order(event: Event, **kwargs):
+    """
+    Ensure:
+        1. There are no S picks before P picks on any station
+        2. There are no amplitude picks before P picks on any station
+    """
+    pdf = obsplus.picks_to_df(event)
+    pdf = pdf.loc[pdf.evaluation_status != "rejected"]
+
+    def pick_order(g, sp, ap, event_id=event.resource_id.id):
+        # get sub dfs with phases of interest
+        p_picks = g[g["phase_hint"].str.upper() == "P"]
+        s_picks = g[g["phase_hint"].str.upper() == "S"]
+        amp_picks = g[g["phase_hint"].str.endswith("AML")]
+        # there should be one P/S pick
+        assert len(p_picks) <= 1 and len(s_picks) <= 1
+        # first check that P is less than S, if not append to name of bad
+        if len(p_picks) and len(s_picks):
+            if s_picks.iloc[0]["time"] < p_picks.iloc[0]["time"]:
+                sp.append(g.name)
+        # next check all amplitude picks are after P
+        if len(p_picks) and len(amp_picks):
+            ptime = p_picks.iloc[0]["time"]
+            bad_amp_picks = amp_picks[amp_picks["time"] < ptime]
+            ap.extend(list(bad_amp_picks["seed_id"]))
+
+    # Checking that picks are in acceptable order
+    gb, sp, ap = pdf.groupby("station"), [], []
+    gb.apply(pick_order, sp, ap)
+    assert len(sp) == 0, "S pick found before P pick:\n" f"station/s: {sp}"
+    assert len(ap) == 0, "amplitude pick found before P pick:\n" f"seed_id/s: {ap}"
+
+
+@validator("obsplus", Event)
+def check_p_lims(event: Event, p_lim=None, **kwargs):
+    """
+    Check for P picks that aren't within p_lim of the median pick (if provided)
+    """
+    if p_lim is not None:
+        df = obsplus.picks_to_df(event)
+        df = df.loc[(df.evaluation_status != "rejected") & (df.phase_hint == "P")]
+        med = df.time.median()
+        bad = df.loc[abs(df.time - med) > p_lim]
+        assert len(bad) == 0, (
+            "Outlying P pick found:\n"
+            f"event_id: {event.resource_id.id}, "
+            f"seed_id/s: {bad.seed_id.tolist()}"
+        )
+
+
+@validator("obsplus", Event)
+def check_amp_lims(event: Event, amp_lim=None, **kwargs):
```

Comment:
See my comments on #64, we need to be pretty careful about the implementation of this. The `amp_lim` should really be specific to a particular magnitude hint.